tiledb/sm/array_schema/array_schema.cc

```diff
@@ -962,12 +962,20 @@ ArraySchema ArraySchema::deserialize(
   // Load version
   // #TODO Add security validation
   auto version = deserializer.read<uint32_t>();
-  if (!(version <= constants::format_version)) {
-    throw ArraySchemaStatusException(
-        "Failed to deserialize array schema: incompatible format version: "
-        "got " +
-        std::to_string(version) +
-        "; expected <= " + std::to_string(constants::format_version));
+  if constexpr (is_experimental_build) {
+    if (version != constants::format_version) {
+      throw ArraySchemaStatusException(
+          "[deserialize] Array format version (" + std::to_string(version) +
```

Comment:
We should make this more specific. Something along the lines of:

```
"[deserialize] This experimental build only works on format version "
  + std::to_string(constants::format_version)
  + ". Version "
  + std::to_string(version)
  + " is not supported by this build."
```

Reply:
Actually, is this even the desired behavior? I assume this is so that folks that attempt to read an array written by an experimental build get a clear error message on what's happened. The change here would mean that experimental builds will no longer be able to read arrays written by non-experimental builds which seems wrong. I'd think the logic should be something like:

```
if (version > constants::format_version) {
 if constexpr (!is_experimental_build) {
   if ((version & 0x80000000) == 0x80000000) {
     throw error that says something like "This is not an experimental build. Unable to read experimental version ${version}. Max version supported is ${constants::format_version}"
    } 
  }

  throw the normal "version ${version} is newer than the library format version ${version}
}
```

Reply:
I was under the impression that the exact equality is for when both arrays are experimental. I think we want the experimental version to still work with versions that are non-experimental and less than or equal to the base version number. This behavior would be useful for testing format version changes on experimental branches still behave nicely with existing arrays while prevent failure from conflicting experimental changes.

We probably also want to consider better error message when the failure was trying to read an experimental array with a non-experimental version.

Reply:
> I think we want the experimental version to still work with versions that are non-experimental and less than or equal to the base version number.

+1

Reply:
Coming back to this, I'd say the logic as I was thinking would be something like:

```c++
if (is_experimental_version(array_version) && !is_experimental_build) {
    throw a descriptive error informing users they're trying to open an experimental array on a non-experimental build
} else if(array_version > library_version) {
   throw the regular error about the array being built by a newer version of libtiledb
}```